### PR TITLE
docker multiplatform image build

### PR DESCRIPTION
**Pull request by smohan-dw**
The multiplatform build needs to be executed on a dedicated machine. Github multiarch action hits OOM after 7 minutes. 

```docker buildx build --platform linux/amd64,linux/arm64 -f docker/Dockerfile -t dhiway/cord-multiarch:test . --push``` 

docker pull dhiway/cord-multiarch:test is the test image built using this dockerfile. 

Also, the final build image is changed to debian:bullseye-slim, as distroless arm64 missing a system library. We need to track this issue and change to distroless once it is fixed.

### CODE REVIEW

**Review by amarts — COMMENTED** [review]
LGTM

File: extras/README.md
Comment (amarts) [inline extras/README.md:1]: Looks like there is conflict here. May need a rebase.

File: .github/workflows/docker-multiarch.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    env:
+      REGISTRY: docker.io
+      USERNAME: dhiway
+      IMAGE_NAME: ${{ github.event.repository.name }}-multiarch-test
+      COMMIT_SHA: ${{ github.sha }}
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
```
Comment (roynrishingha) [inline .github/workflows/docker-multiarch.yml:21]: We need `QEMU` before `buildx`. Please add this:
```
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v2
```

File: .github/workflows/docker-multiarch.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
```
Comment (roynrishingha) [inline .github/workflows/docker-multiarch.yml]: Use this version 2
```
- name: Set up Buildx
  uses: docker/setup-buildx-action@v2
```

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# ------------------------------------------------------------------------------
+# Build Stage
+# ------------------------------------------------------------------------------
+
+FROM rust:1-bullseye as build
+
+LABEL maintainer="engineering@dhiway.com"
+
+ARG target
+ARG profile=release
```
Comment (roynrishingha) [inline docker/Dockerfile:10]: ARG PROFILE=production

Is not it better to make it all caps?

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# ------------------------------------------------------------------------------
+# Build Stage
+# ------------------------------------------------------------------------------
+
+FROM rust:1-bullseye as build
```
Comment (roynrishingha) [inline docker/Dockerfile]: FROM rust:1-bullseye as builder

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+LABEL maintainer="engineering@dhiway.com"
+
+ARG target
+ARG profile=release
+
+WORKDIR /builder
```
Comment (roynrishingha) [inline docker/Dockerfile]: WORKDIR /build

File: docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+# Conditional installation of cross-compilation tools
+RUN if [ "$target" = "linux/arm64" ] || [ "$target" = "linux/arm64/v8" ]; then \
+  apt-get install -y gcc-aarch64-linux-gnu binutils-aarch64-linux-gnu g++-aarch64-linux-gnu && \
+  echo "export CC=aarch64-linux-gnu-gcc" >> /builder/dynenv && \
+  echo "export CXX=aarch64-linux-gnu-g++" >> /builder/dynenv; \
+  fi
+
+# Copy source tree
+COPY . /builder
```
Comment (roynrishingha) [inline docker/Dockerfile]: COPY . /build

File: docker/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+# Configure build environment and build
+RUN set -x && \
+  if [ -f /builder/dynenv ]; then . /builder/dynenv; fi && \
```
Comment (roynrishingha) [inline docker/Dockerfile]: if [ -f /build/dynenv ]; then . /build/dynenv; fi && \

File: docker/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  cargo build --locked --profile $profile && \
+  mkdir -p build && \
+  mv target/$target/$profile/cord build/
+
+# ------------------------------------------------------------------------------
+# Final Stage
+# ------------------------------------------------------------------------------
+
+FROM debian:bullseye-slim
```
Comment (roynrishingha) [inline docker/Dockerfile:39]: Why not use alpine ?

File: docker/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+LABEL maintainer="engineering@dhiway.com"
+LABEL description="CORD Blockchain Node"
+
+COPY --from=build /builder/build/cord /cord
```
Comment (roynrishingha) [inline docker/Dockerfile]: Lets copy in `usr/local/bin` to directly use the binary.

File: docker/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+RUN ["/cord","--version"]
```
Comment (roynrishingha) [inline docker/Dockerfile]: RUN ["cord","--version"]

File: docker/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+USER 1000:1000
+EXPOSE 30333 9933 9944 9615
+
+ENTRYPOINT ["/cord"]
```
Comment (roynrishingha) [inline docker/Dockerfile]: ENTRYPOINT ["cord"]

File: docker/Dockerfile (hunk 8 above)
Comment (smohan-dw) [reply docker/Dockerfile:39]: This is a temporary switch. Working with the distroless team to add the support we need

File: docker/Dockerfile (hunk 9 above)
Comment (smohan-dw) [reply docker/Dockerfile]: not required as this is a temp switch

File: docker/Dockerfile (hunk 10 above)
Comment (smohan-dw) [reply docker/Dockerfile]: not required as this is a temp switch

File: docker/Dockerfile (hunk 11 above)
Comment (smohan-dw) [reply docker/Dockerfile]: not required as this is a temp switch

File: docker/Dockerfile (hunk 3 above)
Comment (smohan-dw) [reply docker/Dockerfile:10]: It's good practice to keep the instructions in CAPS and arguments small.

File: docker/Dockerfile (hunk 4 above)
Comment (smohan-dw) [reply docker/Dockerfile]: Yes, ci-linux is only available for AMD64. The image we are using is a multi-arch image from the Rust project.

File: docker/Dockerfile (hunk 5 above)
Comment (smohan-dw) [reply docker/Dockerfile]: Any reason? I am using a build dir later. might cause confusion

File: .github/workflows/docker-multiarch.yml (hunk 1 above)
Comment (smohan-dw) [reply .github/workflows/docker-multiarch.yml:21]: done, thanks for the catch

File: extras/README.md
Comment (smohan-dw) [reply extras/README.md:1]: done

File: .github/workflows/docker-multiarch.yml (hunk 2 above)
Comment (smohan-dw) [reply .github/workflows/docker-multiarch.yml]: done, thanks for the catch

File: docker/Dockerfile (hunk 5 above)
Comment (roynrishingha) [reply docker/Dockerfile]: No specific reason. `/build` is mostly used, so I mentioned.

**Comment by smohan-dw** [comment]
updated Dockerfile

File: docker/Dockerfile (hunk 7 above)
Comment (smohan-dw) [reply docker/Dockerfile]: changed

File: .github/workflows/docker-multiarch.yml [hunk 12]
```diff
@@ -0,0 +1,64 @@
+name: docker
+
+on:
+  push:
+    branches:
+      - develop
```
Comment (amarts) [inline .github/workflows/docker-multiarch.yml:6]: Can we keep the branch name as something unique (call it 'only-demo' or something) so it doesn't get triggered in github CI, but we use the logic on a builder in our infra to push it to DOCKER? With that, this is good to get in to repo.

**Review by amarts — APPROVED** [review]

File: .github/workflows/docker-multiarch.yml (hunk 12 above)
Comment (smohan-dw) [reply .github/workflows/docker-multiarch.yml:6]: We can keep the docker actions as a reference. Post-merge, these actions (multiarch, docker) can be disabled at the repository level.

**Pull request merged**